grouping_flutter/lib/service/auth/mobile_oauth2.dart

```diff
@@ -0,0 +1,246 @@
+          mode: encryptP.AESMode.ecb,
+          padding: null))
+      .encrypt(_stateCode, iv: encryptP.IV.fromSecureRandom(16));
+
+  _getSignInGrant() {
+    if (pkceSupported) {
+      grant = oauth2.AuthorizationCodeGrant(
+          clientId, authorizationEndpoint, tokenEndpoint,
+          secret: clientSecret,
+          httpClient: JsonFormatHttpClient(),
+          codeVerifier: _pkcePair.codeVerifier);
+      // debugPrint("Pkcepair Verifier: ${_pkcePair.codeVerifier}");
+    } else {
+      grant = oauth2.AuthorizationCodeGrant(
+          clientId, authorizationEndpoint, tokenEndpoint,
+          secret: clientSecret, httpClient: JsonFormatHttpClient());
+    }
+  }
+
+  Future _informVerifierToBackend() async {
+    String stringUrl;
+
+    if (kIsWeb) {
+      stringUrl = '${Config.baseUriWeb}/auth/verifier/';
+    } else {
+      stringUrl = '${Config.baseUriMobile}/auth/verifier/';
+    }
+
+    if (pkceSupported) {
+      // debugPrint(pkceSupported.toString());
+      Response response = await post(Uri.parse(stringUrl),
+          body: {'verifier': pkcePairVerifier.base64});
+    }
+  }
+
+  Future _informStateToBackend() async {
+    String stringUrl;
+
+    if (kIsWeb) {
+      stringUrl = '${Config.baseUriWeb}/auth/state/';
+    } else {
+      stringUrl = '${Config.baseUriMobile}/auth/state/';
+    }
+
+    if (stateSupported) {
+      // debugPrint(stateSupported.toString());
+      debugPrint("stateCode: $_stateCode");
+      Response response = await post(Uri.parse(stringUrl),
+          body: {'state': statePairVerifier.base64});
+    }
+  }
+
+  Future _informPlatform() async {
+    String stringUrl;
+
+    if (kIsWeb) {
+      stringUrl = '${Config.baseUriWeb}/auth/platform/';
+    } else {
+      stringUrl = '${Config.baseUriMobile}/auth/platform/';
+    }
+
+    Response response = await post(Uri.parse(stringUrl),
+        body: {'platform': kIsWeb ? 'web' : 'mobile'});
+  }
+
+  Future _informCode() async {
+    String stringUrl;
+    if (kIsWeb) {
+      stringUrl = '${Config.baseUriWeb}/auth/callback/';
+    } else {
+      stringUrl = '${Config.baseUriMobile}/auth/callback/';
+    }
+    FlutterSecureStorage storage = FlutterSecureStorage();
+
+    String? code = await storage.read(key: 'code');
+    debugPrint(code);
+
+    await get(Uri.parse(stringUrl).replace(queryParameters: {'code': code!}));
+  }
+
+  Future initialLoginFlow() async {
+    _getSignInGrant();
+    if (stateSupported) {
+      authorizationUrl = grant.getAuthorizationUrl(redirectedUrl,
+          scopes: scopes, state: _stateCode);
+    } else {
+      authorizationUrl =
+          grant.getAuthorizationUrl(redirectedUrl, scopes: scopes);
+    }
+    await _informPlatform();
+    await _informVerifierToBackend();
+    await _informStateToBackend();
+    grant.close();
+  }
+
+  // TODO: this is view, no context here
+  Future showWindowAndListen(BuildContext context) async {
```

Comment:
Simple way to solve this problem just pass callback parameter `onError` `onURlChange` `onGrantClose` to avoid pass Build context directly in.